pilot/pkg/serviceregistry/kube/controller/network.go

```diff
@@ -252,23 +250,19 @@ func (c *Controller) extractGatewaysFromService(svc *model.Service) bool {
 }
 
 // reloadNetworkGateways performs extractGatewaysFromService for all services registered with the controller.
+// It returns whether there is a gateway changed.
 // It is called only by `onNetworkChange`.
 // It iterates over all services, because mesh networks can be set with a service name.
-func (c *Controller) reloadNetworkGateways() {
+func (c *Controller) reloadNetworkGateways() bool {
 	c.Lock()
-	gwsChanged := false
+	defer c.Unlock()
+	gwChanged := false
 	for _, svc := range c.servicesMap {
 		if c.extractGatewaysInner(svc) {
-			gwsChanged = true
-			break
```

Comment:
This is a bug in case there are more than one gateway services are specified in the meshNetworks

Reply:
See the below tests updated